Topics/Software_Tools/Git/Git_Merging_and_Rebasing.md

```diff
@@ -11,4 +11,186 @@ Both `git merge` and `git rebase` are methods for integrating new commits that a
+
+One of the most useful features that Git introduced was *branching*. This is the cornerstone of how collaboration is done in Git, but one of the complications is when you want to reconcile the changes made in one branch with the work done in another branch. 
+
+You may be familiar with the most common use case: you want to update your branch with the latest changes from `main`.
+
+There are two different ways of combining two branches: merging and rebasing. They work quite differently, and we’ll explore when and why you’d choose one over the other.
+
+## Merging
+
+The first and classic way to combine the work done in two branches is to *merge* the two branches together. Merging is usually the default way to reconcile changes between branches because it preserves commit history (and is thus “safer”), and resolving merge conflicts is (relatively) simple compared to rebasing.
+
+Say you want to merge some new commits from the `main` branch into your `feature` branch.
+
+Your `feature` branch branched off of the `main` branch at some commit. We’ll illustrate the commits in chronological order (so commit B was committed after commit A).
+
+We’ll say that feature branched off of `main` at some commit C.
+
+So, your `main` branch may look like:
+> A -> B -> C -> E -> F
+
+And, your `feature` branch may look like:
+> A -> B -> C -> D -> G
+
+But notice that there are commits in `main` that happened after `feature` was branched off. In other words, we want commits E and F to be imported into the `feature` branch; perhaps they are useful for the work we’re doing in `feature`. 
+
+You run:
+
+```
+git checkout feature
+git merge main
+```
+
+What  this does is merge `main` into `feature`. This firstly creates a merge commit in `feature` which incorporates all of the changes made in `main` but not in `feature` by comparing their *diffs*. Any changes made in `main` but not `feature` are then applied to `feature` via the merge commit.
+
+Then, in the Git log, you’ll see that the commits in these two branches are preserved, but they then converge with the merge commit.
+
+So, `feature` now has both the commits of `main` and itself, so it is “up to date” with `main`:
+```
+Feature branch:
+
+ A -> B -> C -> E -> F -> (merge commit)
+
+             ↳ D -> G -> ⬏
+```
+
+## Rebasing
+
+As you saw, merging does not hide the fact that `feature` and `main` were being worked on in parallel. But, you might have noticed that another way of importing commits E and F into `feature` would be to change the commit that `feature` branched off of `main` from commit C to commit F. 
+
+Recall that the `main` branch looks like:
+> A -> B -> C -> E -> F
+
+And the `feature` branch looks like:
+> A -> B -> C -> D -> G
+
+Merging resulted in a branch that looked like:
+```
+ A -> B -> C -> E -> F -> (merge commit)
+
+             ↳ D -> G -> ⬏
+```
+
+But, wouldn’t it be nice if `feature` could look like:
+> A -> B -> C -> D -> E -> F -> G
```

Comment:
I think the correct order should be `A -> B -> C -> E -> F -> D -> G`

Reply:
That's correct, good catch.